sdk/core/azure-core/src/main/java/com/azure/core/implementation/ByteBufferCollector.java

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import com.azure.core.util.logging.ClientLogger;
+
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+/**
+ * This class offers functionality similar to {@link ByteArrayOutputStream} but instead of consuming byte arrays it
+ * consumes ByteBuffers. This class is optimized to reduce the number of memory copies by directly writing a passed
+ * ByteBuffers data directly into its backing byte array, this differs from handling for {@link ByteArrayOutputStream}
+ * where ByteBuffer data may need to be first copied into a temporary buffer resulting in an extra memory copy.
+ */
+public final class ByteBufferCollector {
+    /*
+     * Start with a default size of 1 KB as this is small enough to be performant while covering most small response
+     * sizes.
+     */
+    private static final int DEFAULT_INITIAL_SIZE = 1024;
+
+    private static final String INVALID_INITIAL_SIZE = "'initialSize' cannot be equal to or less than 0.";
+    private static final String REQUESTED_BUFFER_INVALID = "Required capacity is greater than Integer.MAX_VALUE.";
+
+    private final ClientLogger logger = new ClientLogger(ByteBufferCollector.class);
+
+    private byte[] buffer;
+    private int position;
+
+    /**
+     * Constructs a new ByteBufferCollector instance with a default sized backing array.
+     */
+    public ByteBufferCollector() {
+        this(DEFAULT_INITIAL_SIZE);
+    }
+
+    /**
+     * Constructs a new ByteBufferCollector instance with a specified initial size.
+     *
+     * @param initialSize The initial size for the backing array.
+     * @throws IllegalArgumentException If {@code initialSize} is equal to or less than {@code 0}.
+     */
+    public ByteBufferCollector(int initialSize) {
+        if (initialSize <= 0) {
+            throw logger.logExceptionAsError(new IllegalArgumentException(INVALID_INITIAL_SIZE));
+        }
+
+        this.buffer = new byte[initialSize];
+        this.position = 0;
+    }
+
+    /**
+     * Writes a ByteBuffers content into the backing array.
+     *
+     * @param byteBuffer The ByteBuffer to concatenate into the collector.
+     */
+    public synchronized void write(ByteBuffer byteBuffer) {
```

Comment:
This is doing some important capacity manipulation, we should have some unit test to validate our logic. For example case when ByteBuffer is of MAX int size.

Reply:
We handle that in the call to ensure the capacity, if ByteBuffer.remaining + buffer.length > Integer.MAX_VALUE we throw an IllegalStateException.